src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -545,7 +545,23 @@ internal static string GetPipeNameForPathOpt(string compilerExeDirectory)
                 return null;
             }
 
-            return $"{userName}.{(isAdmin ? 'T' : 'F')}.{basePipeName}";
+            var pipeName = $"{userName}.{(isAdmin ? 'T' : 'F')}.{basePipeName}";
+
+#if NETCOREAPP2_0
+            if (!PlatformInformation.IsWindows)
+            {
+                // Check if the pipe's path would be longer than a legal path on Mac
+                var getPipePath = typeof(PipeStream).GetMethod("GetPipePath", BindingFlags.Static | BindingFlags.NonPublic);
+                var pipePath = (string)getPipePath.Invoke(null, new[] { ".", pipeName });
```

Comment:
The use of reflection here needs to be wrapped in a `try / catch`. This can change in the future and we don't want to break the compilation pipeline if it does. Probably should factor this out into a new method / local function.